core/dbt/adapters/sql/connections.py

```diff
@@ -152,6 +151,18 @@ def execute(
             table = dbt.clients.agate_helper.empty_table()
         return response, table
 
+    def dry_run(self, sql: str) -> AdapterResponse:
+        """Submit the given SQL to the engine for validation, but not execution.
+
+        By default we simply prefix the query with the explain keyword and allow the
+        exceptions thrown by the underlying engine on invalid SQL inputs to bubble up
+        to the exception handler.
+
+        :param sql str: The sql to validate
+        """
+        explain_sql = f"explain {sql}"
```

Comment:
Generally we would wrap this in a Jinja macro, and call that macro here (via `execute_macro`) instead of Python f-string. Something like:
```sql
{% macro get_explain_sql(query) %}
    explain {{ query }}
{% endmacro %}
```
That way it's "user space" code that someone could could override if they wanted to. Including an adapter maintainer, if the only thing that's different on their data platform is (e.g.) a synonym for `explain`.

There are legitimate cases for using Python templating over Jinja templating: Easier to write, to unit-test. In cases where we believe it's tricky-to-write & functionally uncontroversial, such that no user would realistically want/need to override, that approach can make more sense.

Not saying we need to go one way or the other — just offering my perspective & context on how we've done this in the past.